src/lib/structures/game/BinbGame.ts

```diff
@@ -0,0 +1,172 @@
+import { createEmbed } from '#utils/responses';
+import { container } from '@sapphire/framework';
+import { getDuplicates } from '#utils/common';
+import { isDMChannel } from '@sapphire/discord.js-utilities';
+
+export class BinbGame extends Game {
+	public readonly gameType = GameType.Binb;
+
+	public async guess(message: Message) {
+		const guess = cleanName(message.content);
+		const guesserID = message.author.id;
+		let guessed: AcceptedAnswer | null = null;
+
+		if (!this.round.songGuessers.includes(guesserID)) {
+			guessed = this.processSongGuess(guess, guesserID) ? AcceptedAnswer.Song : null;
+		}
+
+		let artistGuessed = null;
+		if (!guessed) {
+			artistGuessed = this.processArtistGuess(guess, guesserID);
+			guessed = artistGuessed ? AcceptedAnswer.Artist : null;
+		}
+
+		const promises: Promise<unknown>[] = [];
+
+		if (guessed) {
+			const article = guessed === AcceptedAnswer.Artist && artistGuessed !== this.round.primaryArtist ? 'a featuring' : 'the';
+			const embed = createEmbed(`✅ You guessed ${article} ${guessed}`);
+			promises.push(message.reply({ embeds: [embed] }));
+		}
+
+		if (guessed && this.guessedBoth(guesserID)) {
+			// User has guessed both the primary artist and song
+			const embed = createEmbed(`✅ ${message.author.tag} got it in ${((Date.now() - this.round.startTime) / 1000).toPrecision(2)}s!`);
+			promises.push.apply(this.dmAllPlayers({ embeds: [embed] }));
+			if (this.everyoneGuessedBoth()) {
+				promises.push(this.queue.player.stop());
+			}
+		}
+
+		await Promise.all(promises);
+	}
+
+	public async onTrackEnd() {
+		const everybodyPassed = this.round.passedPlayers.size === this.players.size;
+		const { tracksPlayed, playlistLength, nowPlaying } = this.queue;
+		let embedFooter = `${tracksPlayed}/${playlistLength}`;
+		let embedDescription = everybodyPassed ? 'Everybody passed 🏃' : '';
+		// const prevLeaderboard = new Collection(this.leaderboard);
+
+		if (!everybodyPassed) {
+			const { songGuessers, primaryArtistGuessers, artistGuessers, primaryArtist } = this.round;
+			const primaryArtistAndSongGuessers = [...songGuessers, ...primaryArtistGuessers];
+
+			// Give 0.5 points for each of primary artist and song
+			for (const guesser of primaryArtistAndSongGuessers) {
+				this.leaderboard.inc(guesser, 0.5);
+			}
+
+			// Give 0.25 points for each 'other' artist
+			for (const [artist, guessers] of artistGuessers.entries()) {
+				// Skip primary artist because it's already scored
+				if (artist === primaryArtist) {
+					continue;
+				}
+
+				for (const guesser of guessers) {
+					this.leaderboard.inc(guesser, 0.25);
+				}
+			}
+
+			// Give bonus points to 1st, 2nd, 3rd double guessers
+			const doubleGuessers = getDuplicates(primaryArtistAndSongGuessers);
+			for (const [index, doubleGuesser] of doubleGuessers.entries()) {
+				this.leaderboard.inc(doubleGuesser, 1.5 - index * 0.5);
+				if (index === 2) {
+					break;
+				}
+			}
```

Comment:
This would be ~75% faster if there were at least 3 double guessers, but only ~2% slower if there weren't
```suggestion
			// Cut back on unnecessary loops if they're aren't more than 3 double guessers.
			const placedSize = Math.min(3, doubleGuessers.length);
			for (let i = 0; i < placedSize; i++) {
				this.leaderboard.inc(doubleGuessers[i], 1.5 - i * 0.5);
			}
```